--- a/drivers/input/joystick/xpad.c	2025-03-20 16:23:02.661181765 +0300
+++ a/drivers/input/joystick/xpad.c	2025-03-24 15:44:38.785781654 +0300
@@ -2319,22 +2319,26 @@
 	struct input_dev *input = xpad->dev;
 	int retval = 0;
 
-	if (xpad->xtype == XTYPE_XBOX360W) {
-		retval = xpad360w_start_input(xpad);
-	} else {
-		mutex_lock(&input->mutex);
-		if (input_device_enabled(input)) {
-			retval = xpad_start_input(xpad);
-		} else if (xpad->xtype == XTYPE_XBOXONE) {
-			/*
-			 * Even if there are no users, we'll send Xbox One pads
-			 * the startup sequence so they don't sit there and
-			 * blink until somebody opens the input device again.
-			 */
-			retval = xpad_start_xbox_one(xpad);
-		}
+//
+	if (xpad->xtype == XTYPE_XBOX360W)
+		return xpad360w_start_input(xpad);
+
+	mutex_lock(&input->mutex);
+
+	if (input_device_enabled(input)) {
+		retval = xpad_start_input(xpad);
 		mutex_unlock(&input->mutex);
+		return retval;  }
+
+	if (xpad->xtype == XTYPE_XBOXONE) {
+		/*
+		 * Even if there are no users, we'll send Xbox One pads
+		 * the startup sequence so they don't sit there and
+		 * blink until somebody opens the input device again.
+		 */
+		retval = xpad_start_xbox_one(xpad);
 	}
+	mutex_unlock(&input->mutex);
 
 	return retval;
 }
