--- a/drivers/hid/hid-debug.c	2026-01-08 18:40:53.430184196 +0300
+++ a/drivers/hid/hid-debug.c	2026-01-09 19:24:22.592441824 +0300
@@ -967,6 +967,7 @@
 	[ABS_DISTANCE] = "Distance",	[ABS_TILT_X] = "XTilt",
 	[ABS_TILT_Y] = "YTilt",		[ABS_TOOL_WIDTH] = "ToolWidth",
 	[ABS_VOLUME] = "Volume",	[ABS_MISC] = "Misc",
+	[ABS_PROFILE] = "Profile",
 	[ABS_MT_TOUCH_MAJOR] = "MTMajor",
 	[ABS_MT_TOUCH_MINOR] = "MTMinor",
 	[ABS_MT_WIDTH_MAJOR] = "MTMajorW",
